src/main/java/io/split/android/client/service/sseclient/reactor/SplitUpdatesWorker.java

```diff
@@ -18,24 +28,131 @@ public class SplitUpdatesWorker extends UpdateWorker {
+    private final CompressionUtilProvider mCompressionUtilProvider;
+    private final SplitTaskExecutor mSplitTaskExecutor;
+    private final SplitTaskFactory mSplitTaskFactory;
+    private final Base64Decoder mBase64Decoder;
 
     public SplitUpdatesWorker(@NonNull Synchronizer synchronizer,
-                              @NonNull BlockingQueue<SplitsChangeNotification> notificationsQueue) {
+                              @NonNull BlockingQueue<SplitsChangeNotification> notificationsQueue,
+                              @NonNull SplitsStorage splitsStorage,
+                              @NonNull CompressionUtilProvider compressionUtilProvider,
+                              @NonNull SplitTaskExecutor splitTaskExecutor,
+                              @NonNull SplitTaskFactory splitTaskFactory) {
+        this(synchronizer,
+                notificationsQueue,
+                splitsStorage,
+                compressionUtilProvider,
+                splitTaskExecutor,
+                splitTaskFactory,
+                new Base64DecoderImpl());
+    }
+
+    @VisibleForTesting
+    public SplitUpdatesWorker(@NonNull Synchronizer synchronizer,
+                              @NonNull BlockingQueue<SplitsChangeNotification> notificationsQueue,
+                              @NonNull SplitsStorage splitsStorage,
+                              @NonNull CompressionUtilProvider compressionUtilProvider,
+                              @NonNull SplitTaskExecutor splitTaskExecutor,
+                              @NonNull SplitTaskFactory splitTaskFactory,
+                              @NonNull Base64Decoder base64Decoder) {
         super();
         mSynchronizer = checkNotNull(synchronizer);
         mNotificationsQueue = checkNotNull(notificationsQueue);
+        mSplitsStorage = checkNotNull(splitsStorage);
+        mCompressionUtilProvider = checkNotNull(compressionUtilProvider);
+        mSplitTaskExecutor = checkNotNull(splitTaskExecutor);
+        mSplitTaskFactory = checkNotNull(splitTaskFactory);
+        mBase64Decoder = checkNotNull(base64Decoder);
     }
 
     @Override
     protected void onWaitForNotificationLoop() throws InterruptedException {
         try {
             SplitsChangeNotification notification = mNotificationsQueue.take();
-            mSynchronizer.synchronizeSplits(notification.getChangeNumber());
-            Logger.d("A new notification to update splits has been received. " +
-                    "Enqueuing polling task.");
+            Logger.d("A new notification to update splits has been received");
+
+            long storageChangeNumber = mSplitsStorage.getTill();
+            if (notification.getChangeNumber() <= storageChangeNumber) {
+                Logger.d("Notification change number is lower than the current one. Ignoring notification");
+                return;
+            }
+
+            if (isLegacyNotification(notification) || isInvalidChangeNumber(notification, storageChangeNumber)) {
```

Comment:
please ping me to chat about this implementation. we must to make sure maintain backward compatibility